### PR TITLE
feat: add formats prop

### PR DESCRIPTION
**Pull request by samydoesit**
Add formats prop

#356

### CODE REVIEW

**Comment by gruhn** [comment]
Amazing, thanks! Could you also add a demo that shows the feature in action on? Maybe on `QrcodeStream` and with a (multi-)select field, where users can select the barcode formats to be used. The [Visual Tracking demo](https://vue-qrcode-reader.netlify.app/demos/CustomTracking.html) has a similar setup. Maybe it's also useful to turn on tracking for the new demo so we can see how tracking behaves on the different barcode formats.

**Comment by samydoesit** [comment]
Duplicated the "Visual Tracking" Page added "Visual Tracking with Formats" (though the naming might not be ideal).

**Comment by gruhn** [comment]
> Regarding the component props reactivity ... 

Exactly, this is the tricky part. 

> I attempted to add props.formats to the cameraSettings computed property in QrcodeStream, so that the cameraSettings watcher also responds to changes when the formats change.

I think it's not necessary to reload the camera when `formats` changes. It should be possible to let the camera stream running but just switch the `BarcodeDetector` instance "on-the-fly". Maybe we can just export a `setBarcodeDetector` function from `scanner.ts` so the instance can be replaced during scanning. But I'm not sure if that's the most elegant approach or if there might be problems with that.

Incidentally, I think this check:
```
if(cameraSettings.formats !== prevCameraSettings.formats) {
```
might be true too often, since it checks equality by reference. For example: `["qr_code"] !== ["qr_code"]` is `true` because the objects on both sides have different references.

**Comment by samydoesit** [comment]
I tried to avoid introducing a side effect, but replacing the BarcodeDetector instance seemed to work well when I tested it.

**Comment by gruhn** [comment]
> I tried to avoid introducing a side effect

I also try to avoid that most the time and rely on the reactive system as much as possible. But as I said, not sure what would be nicer here. 

> seemed to work well when I tested it.

Same here good job 👍

File: src/misc/scanner.ts [hunk 1]
```diff
@@ -11,11 +17,13 @@ export const keepScanning = async (
...
+    formats = ['qr_code'],
     minDelay
-  }: { detectHandler: any; locateHandler: any; minDelay: any }
+  }: { detectHandler: any; locateHandler: any; minDelay: any, formats: BarcodeFormat[] }
 ) => {
-  const barcodeDetector = new BarcodeDetector({ formats: ['qr_code'] })
-
+  if(!barcodeDetector) {
+    setScanningFormats(formats)
+  }
```
Comment (gruhn) [inline src/misc/scanner.ts:26]: Here we might have the situation that the BarcodeDetector instance is preserved over the lifetime of the QrcodeStream component. For example what if we create:
```
<qrcode-stream :formats="['qr_code']" />
```
Do some scanning, destroy the component, and then create a new 
```
<qrcode-stream :formats="['code_93']" />
```
Then I think the BarcodeDetector instance is still alive and we don't create a new one, so we keep scanning with formats `['qr_code']`.

Comment (gruhn) [reply src/misc/scanner.ts]: sounds good :+1:

File: docs/.vitepress/components/demos/CustomTrackingWithFormats.vue [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+      { text: 'outline', value: this.paintOutline },
+      { text: 'centered text', value: this.paintCenterText },
+      { text: 'bounding box', value: this.paintBoundingBox }
+    ]
+
+    const selected = options[1]
+
+    const barcodeTypes = ["aztec", "code_128", "code_39", "code_93", "codabar", "data_matrix", "ean_13", "ean_8", "itf", "pdf417", "qr_code", "upc_a", "upc_e"]
+
+    const selectedBarcodeType = barcodeTypes[1]
```
Comment (gruhn) [inline docs/.vitepress/components/demos/CustomTrackingWithFormats.vue:42]: On the other hand this is not an array.

**Comment by gruhn** [comment]
fyi: here are examples images for each barcode types: https://github.com/gruhn/barcode-detector/tree/master/example_codes

File: docs/.vitepress/components/demos/CustomTrackingWithFormats.vue [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    <p>
+      Track function:
+      <select v-model="selected">
+        <option v-for="option in options" :key="option.text" :value="option">
+          {{ option.text }}
+        </option>
+      </select>
+    </p>
+
+    <qrcode-stream :formats="[selectedBarcodeType]" :track="selected.value" @error="logErrors" />
```
[the remark this answers is not in the document]
Comment (samydoesit) [reply docs/.vitepress/components/demos/CustomTrackingWithFormats.vue]: Oops missed that 🙈

File: docs/.vitepress/components/demos/CustomTrackingWithFormats.vue (hunk 2 above)
Comment (samydoesit) [reply docs/.vitepress/components/demos/CustomTrackingWithFormats.vue]: Missed that as well. It's fixed now

**Comment by gruhn** [comment]
Alright, looks good to me :slightly_smiling_face:

**Pull request merged**